---
title: "biocXqtl: flexible molecular QTL assessment"
author: "Vincent J. Carey, stvjc at channing.harvard.edu"
date: "`r format(Sys.time(), '%B %d, %Y')`"
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{biocXqtl: flexible molecular QTL assessment}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::html_document:
    highlight: pygments
    number_sections: yes
    theme: united
    toc: yes
---

# Introduction

eQTL analysis generally involves large data volumes
of genotype calls.  No standard representation
for genotypes has emerged, though VCFs and plink2
files are frequently encountered.

In this package we provide a flexible approach to
estimating associations between genotype and
molecular phenotypes such as gene expression
or DNA methylation.  We will use the
RangedSummarizedExperiment structure as the
primary element for structuring the inputs
to association test procedures.


# Working with variants from VCF

Thanks to the AWS Open Data project, tabix-indexed VCFs for genotyped
lymphoblastoid cell lines from the 1000 genomes project
are [available](https://registry.opendata.aws/1000-genomes/).
Data from the Multi-Ancestry analysis of Gene Expression
are [available](https://github.com/mccoy-lab/MAGE)
via Zenodo or
Dropbox.  In our experience the Dropbox link is more performant.

We have acquired the "DESeq2" gene quantifications from
the MAGE archive, and a selection of `release` genotypes
from the associated cell lines.

## Data on expression and genotype

The expression data:
```{r theex, message=FALSE}
library(biocXqtl)
data(mageSE_19)
```

Genotype data:
```{r lkgt}
dv = demo_vcf()
h = VariantAnnotation::scanVcfHeader(dv)
head(samples(h))
length(intersect(samples(h), colnames(mageSE_19)))
```

## Building an XqtlExperiment

We will focus on SNVs but the genotype information
has other types of variants.

Bind the minor allele counts to the expression data:
```{r dobind}
mins  = minorAlleleCounts(demo_vcf(), GRanges("19:1-50000000"))
mxx = XqtlExperiment(mageSE_19, mins)
colData(mxx) = NULL
```
The last command above allows us to compute crude measures of association.
If we populate colData with covariate information, test procedures
in the package will incorporate it.

## Creating association statistics with visualizations

The following step uses C++ modules to compute association tests for
all genotypes and all gene expression measures in mageSE_19.
```{r dotests}
sds = rowSds(assay(mxx), na.rm=TRUE)
qq = quantile(sds, .95)
ok = which(sds > qq)
system.time(zzz <- bind_Zs(mxx[ok,]))
```

Here's a helper function to visualize one association.
```{r helper}
onebox = function(xse, mfeat="ENSG00000174837", vnt="rs4897932", title) {
if (missing(title)) title=""
boxplot(split(as.numeric(assay(xse[mfeat,])),
   as.numeric(data.matrix(mcols(getCalls(xse)[vnt,])))),
   ylab=mfeat, xlab=vnt, main=title)
}
onebox(mxx[ok,], title="MAGE eQTL")
```

We can also visualize interactively:
```{r doviz2}
viz_stats(zzz, midchop=5)
```

## Covariate adjustment

First we reanalyze with adjustment for continental group.

```{r redo}
data(mageSE_19)
sds = rowSds(assay(mageSE_19), na.rm=TRUE)
qq = quantile(sds, .9)
ok = which(sds > qq)
cd = colData(mageSE_19)
mins  = minorAlleleCounts(demo_vcf(), GRanges("19:1-3000000"))
mxx = XqtlExperiment(mageSE_19[ok,], mins)
print(mxx)
colData(mxx)=NULL
rowRanges(mxx) = rowRanges(mxx)[,1:6]
mxx$continent = cd$continentalGroup
system.time(zzz <- bind_Zs(mxx))
viz_stats(zzz, midchop=5)
```

Then we add sex as well.

```{r redo2}
data(mageSE_19)
sds = rowSds(assay(mageSE_19), na.rm=TRUE)
qq = quantile(sds, .9)
ok = which(sds > qq)
cd = colData(mageSE_19)
mins  = minorAlleleCounts(demo_vcf(), GRanges("19:1-3000000"))
mxx = XqtlExperiment(mageSE_19[ok,], mins)
mxx$continent = cd$continentalGroup
mxx$sex = cd$sex
system.time(zzz <- bind_Zs(mxx))
viz_stats(zzz, midchop=7)
```

