---
title: "Handling variants in biocXqtl"
shorttitle: "outlier dytection methods"
author: "Vincent J. Carey, stvjc at channing.harvard.edu"
date: "`r format(Sys.time(), '%B %d, %Y')`"
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{Handling variants in biocXqtl}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::html_document:
    highlight: pygments
    number_sections: yes
    theme: united
    toc: yes
---

# Introduction

Diverse strategies have been proposed for managing archives of DNA variants in humans.

- [VCF](https://en.wikipedia.org/wiki/Variant_Call_Format) is widely used and many toolsets are available
for working with this format.  
- [Plink](https://www.cog-genomics.org/plink/2.0/) has a format in wide use.
- Hail's MatrixTable and
associated tooling offer
["easy ways to slice, dice, query, and summarize"](https://hail.is/docs/0.2/tutorials/01-genome-wide-association-study.html) DNA variants.


We'll focus on VCF and Plink in this document.

# VCF access and manipulations

We will work with bgzipped, tabix-indexed resources.

A demonstration VCF archive is available with biocXqtl.
The associated vcf.gz file was formed by filtering publicly
available variants from the 1000 genomes project.
See the Appendix for information on how the demonstration
VCF was produced.  A key filtering step involved limiting
inclusion to those variants for which at least 10 individuals
were present in each of the classes: homozygous rare, heterozygous,
homozygous common.

Setup:
```{r lksn1, message=FALSE}
library(biocXqtl)
library(VariantAnnotation)
vpath = demo_vcf()
```

First we examine the header.

```{r lkh}
h = scanVcfHeader(vpath)
h
```

The header report indicates that genotypes are available for 731 individuals.

Direct ingestion of VCF can consume considerable RAM.  We can select
regions of interest and scan with inline filtering.  See the documentation
on `ScanVcfParam` in VariantAnnotation.

```{r doscan}
library(GenomicRanges)
p = GRanges("19:1-3000000")
ss = ScanVcfParam(which=p)
sc1 = scanVcf(vpath, param=ss)
```

Each range in the `ScanVcfParam` will produce a list element
in the result of `scanVcf`.

```{r lkscan}
names(sc1[[1]])
dim(sc1[[1]]$GENO[[1]])
sc1[[1]]$rowRanges
```

For purposes of statistical modeling, we have a helper
function that retrieves minor allele counts from a tabix-indexed VCF.
This returns a GRanges with counts in the mcols.  The
addresses are bound with the counts.

```{r getmac}
cnts = minorAlleleCounts(vpath, region=GRanges("19:1-3000000"))
cnts[,1:2]
```

# PLINK

## Example data

The data were extracted from the GWAS tutorial.

```{r lkdpl, message=FALSE}
zfi = system.file("plink", "chr19_1kglim.zip", package="biocXqtl")
tdir = tempdir()
unzip(zfi, exdir=tdir)
dir(tdir)
```

## Using genio

This package seems to be capable only of 'full reads'.

```{r lkgenio, message=FALSE, eval=FALSE}
library(genio)
fi = grep("bed$", dir(tdir, full=TRUE), value=TRUE)
dem = read_plink(fi)
dim(dem$X)
```

## Using python `bed-reader`

Here we can make targeted reads from the resource.

```{r dopy, message=FALSE}
library(reticulate)
py_require("bed-reader")
br = import("bed_reader")
fi = grep("bed$", dir(tdir, full=TRUE), value=TRUE)
res = br$open_bed(fi)
toget = tuple(0:4, 0:5)
limread = res$read(toget)
dim(limread)
head(res$bp_position)
head(res$chromosome)
head(res$sid)
head(res$iid)
```

# Appendix

## VCF production

Code to produce the demonstration VCF is

```
library(biocXqtl)
library(VariantAnnotation)
data(mageSE_19) # produced from MAGE "DESeq2" quantifications
cc = colnames(mageSE_19)
data(tiles38)
library(GenomeInfoDb)
seqlevelsStyle(tiles38) = "Ensembl"
t19 = tiles38[which(seqnames(tiles38)=="19")]
bigvcf = filterVcfByGenotypeCounts("~/MAGE/ALL.chr19.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz", samples=cc, min_per_genotype=10, regions=range(t19))
```

The vcf.gz file was obtained following instructions at `https://registry.opendata.aws/1000-genomes/`.

The code given above runs on a 32 GB macbook in a few minutes, with the following report:

```
Total samples in VCF: 2504
Analyzing 731 samples
Reading VCF...
[W::hts_idx_load3] The index file is older than the data file: 
 .../MAGE/ALL.chr19.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz.tbi
Read 1816698 variants
Counting genotypes...
Variants passing filter: 156943 / 1816698
  Mean counts - HomRef: 690.8, Het: 25.4, HomAlt: 14.4
```

## PLINK resource production

Folder `GWAS/project1/data/raw` in the tutorial download has
the relevant contents.

An extract from chr19 is produced using
```
plink2 \
  --pfile all_hg38 vzs \
  --chr 19 \
  --maf 0.1 \
  --keep kp.txt \
  --make-pgen vzs \
  --allow-extra-chr \
  --out chr19_limsamples_pp1
```
`kp.txt` is in `inst/plink` and is a selection of samples
with diverse ancestries.

PLINK bed-format files are produced using
```
plink2 \
  --pfile chr19_limsamples_pp1 vzs \
  --make-bed \
  --out chr19_limsamples_maf10
```
