<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Handling variants in biocXqtl • biocXqtl</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Handling variants in biocXqtl">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">biocXqtl</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.9</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/biocXqtl.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/snps.html">Handling variants in biocXqtl</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/vjcitn/biocXqtl/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Handling variants in biocXqtl</h1>
                        <h4 data-toc-skip class="author">Vincent J. Carey, stvjc at channing.harvard.edu</h4>
            
            <h4 data-toc-skip class="date">December 31, 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/vjcitn/biocXqtl/blob/HEAD/vignettes/snps.Rmd"><code>vignettes/snps.Rmd</code></a></small>
      <div class="d-none name"><code>snps.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Diverse strategies have been proposed for managing archives of DNA variants in humans.</p>
<ul>
<li>
<a href="https://en.wikipedia.org/wiki/Variant_Call_Format" class="external-link">VCF</a> is widely used and many toolsets are available for working with this format.<br>
</li>
<li>
<a href="https://www.cog-genomics.org/plink/2.0/" class="external-link">Plink</a> has a format in wide use.</li>
<li>Hail’s MatrixTable and associated tooling offer <a href="https://hail.is/docs/0.2/tutorials/01-genome-wide-association-study.html" class="external-link">“easy ways to slice, dice, query, and summarize”</a> DNA variants.</li>
</ul>
<p>We’ll focus on VCF and Plink in this document.</p>
</div>
<div class="section level2">
<h2 id="vcf-access-and-manipulations">VCF access and manipulations<a class="anchor" aria-label="anchor" href="#vcf-access-and-manipulations"></a>
</h2>
<p>We will work with bgzipped, tabix-indexed resources.</p>
<p>A demonstration VCF archive is available with biocXqtl. The associated vcf.gz file was formed by filtering publicly available variants from the 1000 genomes project. See the Appendix for information on how the demonstration VCF was produced. A key filtering step involved limiting inclusion to those variants for which at least 10 individuals were present in each of the classes: homozygous rare, heterozygous, homozygous common.</p>
<p>Setup:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vjcitn/biocXqtl">biocXqtl</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: package 'GenomicRanges' was built under R version 4.5.2</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">VariantAnnotation</span><span class="op">)</span></span>
<span><span class="va">vpath</span> <span class="op">=</span> <span class="fu"><a href="../reference/demo_vcf.html">demo_vcf</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>First we examine the header.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/scanVcf-methods.html" class="external-link">scanVcfHeader</a></span><span class="op">(</span><span class="va">vpath</span><span class="op">)</span></span>
<span><span class="va">h</span></span></code></pre></div>
<pre><code><span><span class="co">## class: VCFHeader </span></span>
<span><span class="co">## samples(731): HG00096 HG00100 ... NA21129 NA21130</span></span>
<span><span class="co">## meta(5): fileDate fileformat reference source contig</span></span>
<span><span class="co">## fixed(2): FILTER ALT</span></span>
<span><span class="co">## info(27): CIEND CIPOS ... EX_TARGET MULTI_ALLELIC</span></span>
<span><span class="co">## geno(1): GT</span></span></code></pre>
<p>The header report indicates that genotypes are available for 731 individuals.</p>
<p>Direct ingestion of VCF can consume considerable RAM. We can select regions of interest and scan with inline filtering. See the documentation on <code>ScanVcfParam</code> in VariantAnnotation.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"19:1-3000000"</span><span class="op">)</span></span>
<span><span class="va">ss</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/ScanVcfParam-class.html" class="external-link">ScanVcfParam</a></span><span class="op">(</span>which<span class="op">=</span><span class="va">p</span><span class="op">)</span></span>
<span><span class="va">sc1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/scanVcf-methods.html" class="external-link">scanVcf</a></span><span class="op">(</span><span class="va">vpath</span>, param<span class="op">=</span><span class="va">ss</span><span class="op">)</span></span></code></pre></div>
<p>Each range in the <code>ScanVcfParam</code> will produce a list element in the result of <code>scanVcf</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sc1</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "rowRanges" "REF"       "ALT"       "QUAL"      "FILTER"    "INFO"     </span></span>
<span><span class="co">## [7] "GENO"</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">sc1</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">GENO</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 9594  731</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc1</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">rowRanges</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 9594 ranges and 0 metadata columns:</span></span>
<span><span class="co">##               seqnames          ranges strand</span></span>
<span><span class="co">##                  &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##   rs567986644       19          160565      *</span></span>
<span><span class="co">##   rs548347106       19   183097-183102      *</span></span>
<span><span class="co">##   rs527871033       19   212785-212786      *</span></span>
<span><span class="co">##   rs533243783       19          219620      *</span></span>
<span><span class="co">##    rs62102979       19          226776      *</span></span>
<span><span class="co">##           ...      ...             ...    ...</span></span>
<span><span class="co">##    rs12982139       19         2999553      *</span></span>
<span><span class="co">##   rs199521660       19         2999724      *</span></span>
<span><span class="co">##   rs200315688       19 2999734-2999738      *</span></span>
<span><span class="co">##    rs13345197       19         2999759      *</span></span>
<span><span class="co">##    rs62125443       19         2999804      *</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>For purposes of statistical modeling, we have a helper function that retrieves minor allele counts from a tabix-indexed VCF. This returns a GRanges with counts in the mcols. The addresses are bound with the counts.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cnts</span> <span class="op">=</span> <span class="fu"><a href="../reference/minorAlleleCounts.html">minorAlleleCounts</a></span><span class="op">(</span><span class="va">vpath</span>, region<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"19:1-3000000"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cnts</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 9594 ranges and 2 metadata columns:</span></span>
<span><span class="co">##               seqnames          ranges strand |   HG00096   HG00100</span></span>
<span><span class="co">##                  &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   rs567986644       19          160565      * |         0         1</span></span>
<span><span class="co">##   rs548347106       19   183097-183102      * |         1         1</span></span>
<span><span class="co">##   rs527871033       19   212785-212786      * |         2         1</span></span>
<span><span class="co">##   rs533243783       19          219620      * |         1         1</span></span>
<span><span class="co">##    rs62102979       19          226776      * |         1         1</span></span>
<span><span class="co">##           ...      ...             ...    ... .       ...       ...</span></span>
<span><span class="co">##    rs12982139       19         2999553      * |         1         0</span></span>
<span><span class="co">##   rs199521660       19         2999724      * |         0         1</span></span>
<span><span class="co">##   rs200315688       19 2999734-2999738      * |         1         0</span></span>
<span><span class="co">##    rs13345197       19         2999759      * |         1         0</span></span>
<span><span class="co">##    rs62125443       19         2999804      * |         0         2</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from GRCh38 genome; no seqlengths</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="plink">PLINK<a class="anchor" aria-label="anchor" href="#plink"></a>
</h2>
<div class="section level3">
<h3 id="example-data">Example data<a class="anchor" aria-label="anchor" href="#example-data"></a>
</h3>
<p>The data were extracted from the GWAS tutorial.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">zfi</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"plink"</span>, <span class="st">"chr19_1kglim.zip"</span>, package<span class="op">=</span><span class="st">"biocXqtl"</span><span class="op">)</span></span>
<span><span class="va">tdir</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html" class="external-link">unzip</a></span><span class="op">(</span><span class="va">zfi</span>, exdir<span class="op">=</span><span class="va">tdir</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="va">tdir</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "BiocStyle"                      "chr19_limsamples_maf10.bed"    </span></span>
<span><span class="co">##  [3] "chr19_limsamples_maf10.bim"     "chr19_limsamples_maf10.fam"    </span></span>
<span><span class="co">##  [5] "chr19_limsamples_maf10.log"     "filea8a43da011e3"              </span></span>
<span><span class="co">##  [7] "filea8a441199817"               "filea8a44cdc458d"              </span></span>
<span><span class="co">##  [9] "filea8a458218b25"               "filea8a4739b6aed"              </span></span>
<span><span class="co">## [11] "make_bed.sh"                    "rmarkdown-stra8a44fa9d0a4.html"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="using-genio">Using genio<a class="anchor" aria-label="anchor" href="#using-genio"></a>
</h3>
<p>This package seems to be capable only of ‘full reads’.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/OchoaLab/genio" class="external-link">genio</a></span><span class="op">)</span></span>
<span><span class="va">fi</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"bed$"</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="va">tdir</span>, full<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>, value<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">dem</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/genio/man/read_plink.html" class="external-link">read_plink</a></span><span class="op">(</span><span class="va">fi</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">dem</span><span class="op">$</span><span class="va">X</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 163640    104</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="using-python-bed-reader">Using python <code>bed-reader</code><a class="anchor" aria-label="anchor" href="#using-python-bed-reader"></a>
</h3>
<p>Here we can make targeted reads from the resource.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rstudio.github.io/reticulate/" class="external-link">reticulate</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: package 'reticulate' was built under R version 4.5.2</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/py_require.html" class="external-link">py_require</a></span><span class="op">(</span><span class="st">"bed-reader"</span><span class="op">)</span></span>
<span><span class="va">br</span> <span class="op">=</span> <span class="fu"><a href="https://rstudio.github.io/reticulate/reference/import.html" class="external-link">import</a></span><span class="op">(</span><span class="st">"bed_reader"</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">=</span> <span class="va">br</span><span class="op">$</span><span class="fu">open_bed</span><span class="op">(</span><span class="va">fi</span><span class="op">)</span></span>
<span><span class="va">toget</span> <span class="op">=</span> <span class="fu"><a href="https://rstudio.github.io/reticulate/reference/tuple.html" class="external-link">tuple</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">4</span>, <span class="fl">0</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">limread</span> <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="fu">read</span><span class="op">(</span><span class="va">toget</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">limread</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 5 6</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">bp_position</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 80840 81039 81806 90854 93234 95981</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">chromosome</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "19" "19" "19" "19" "19" "19"</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">sid</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "rs201816663" "rs879919317" "rs2432259"   "rs200254023" "rs151275984"</span></span>
<span><span class="co">## [6] "rs8110113"</span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">iid</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "HG00096" "HG00097" "HG00099" "HG00100" "HG00101" "HG00171"</span></span></code></pre>
<p>genio</p>
</div>
</div>
<div class="section level2">
<h2 id="appendix">Appendix<a class="anchor" aria-label="anchor" href="#appendix"></a>
</h2>
<div class="section level3">
<h3 id="vcf-production">VCF production<a class="anchor" aria-label="anchor" href="#vcf-production"></a>
</h3>
<p>Code to produce the demonstration VCF is</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vjcitn/biocXqtl">biocXqtl</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">VariantAnnotation</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">mageSE_19</span><span class="op">)</span> <span class="co"># produced from MAGE "DESeq2" quantifications</span></span>
<span><span class="va">cc</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mageSE_19</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">tiles38</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomeInfoDb" class="external-link">GenomeInfoDb</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevelsStyle.html" class="external-link">seqlevelsStyle</a></span><span class="op">(</span><span class="va">tiles38</span><span class="op">)</span> <span class="op">=</span> <span class="st">"Ensembl"</span></span>
<span><span class="va">t19</span> <span class="op">=</span> <span class="va">tiles38</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Seqinfo/man/seqinfo.html" class="external-link">seqnames</a></span><span class="op">(</span><span class="va">tiles38</span><span class="op">)</span><span class="op">==</span><span class="st">"19"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">bigvcf</span> <span class="op">=</span> <span class="fu"><a href="../reference/filterVcfByGenotypeCounts.html">filterVcfByGenotypeCounts</a></span><span class="op">(</span><span class="st">"~/MAGE/ALL.chr19.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz"</span>, samples<span class="op">=</span><span class="va">cc</span>, min_per_genotype<span class="op">=</span><span class="fl">10</span>, regions<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">t19</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>The vcf.gz file was obtained following instructions at <code>https://registry.opendata.aws/1000-genomes/</code>.</p>
<p>The code given above runs on a 32 GB macbook in a few minutes, with the following report:</p>
<pre><code>Total samples in VCF: 2504
Analyzing 731 samples
Reading VCF...
[W::hts_idx_load3] The index file is older than the data file: 
 .../MAGE/ALL.chr19.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz.tbi
Read 1816698 variants
Counting genotypes...
Variants passing filter: 156943 / 1816698
  Mean counts - HomRef: 690.8, Het: 25.4, HomAlt: 14.4</code></pre>
</div>
<div class="section level3">
<h3 id="plink-resource-production">PLINK resource production<a class="anchor" aria-label="anchor" href="#plink-resource-production"></a>
</h3>
<p>Folder <code>GWAS/project1/data/raw</code> in the tutorial download has the relevant contents.</p>
<p>An extract from chr19 is produced using</p>
<pre><code>plink2 \
  --pfile all_hg38 vzs \
  --chr 19 \
  --maf 0.1 \
  --keep kp.txt \
  --make-pgen vzs \
  --allow-extra-chr \
  --out chr19_limsamples_pp1</code></pre>
<p><code>kp.txt</code> is in <code>inst/plink</code> and is a selection of samples with diverse ancestries.</p>
<p>PLINK bed-format files are produced using</p>
<pre><code>plink2 \
  --pfile chr19_limsamples_pp1 vzs \
  --make-bed \
  --out chr19_limsamples_maf10</code></pre>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Vince Carey.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
