<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Filter VCF to variants with minimum counts in each genotype group — filterVcfByGenotypeCounts • biocXqtl</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Filter VCF to variants with minimum counts in each genotype group — filterVcfByGenotypeCounts"><meta name="description" content="Filter VCF to variants with minimum counts in each genotype group"><meta property="og:description" content="Filter VCF to variants with minimum counts in each genotype group"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">biocXqtl</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.6</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/biocXqtl.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/vjcitn/biocXqtl/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Filter VCF to variants with minimum counts in each genotype group</h1>
      <small class="dont-index">Source: <a href="https://github.com/vjcitn/biocXqtl/blob/HEAD/R/vcf_utils.R"><code>R/vcf_utils.R</code></a></small>
      <div class="d-none name"><code>filterVcfByGenotypeCounts.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Filter VCF to variants with minimum counts in each genotype group</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">filterVcfByGenotypeCounts</span><span class="op">(</span></span>
<span>  <span class="va">vcf_file</span>,</span>
<span>  samples <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min_per_genotype <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  genome <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span>  regions <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-vcf-file">vcf_file<a class="anchor" aria-label="anchor" href="#arg-vcf-file"></a></dt>
<dd><p>Path to tabix-indexed VCF file (.vcf.gz with .tbi index)</p></dd>


<dt id="arg-samples">samples<a class="anchor" aria-label="anchor" href="#arg-samples"></a></dt>
<dd><p>Character vector of sample IDs to retain</p></dd>


<dt id="arg-min-per-genotype">min_per_genotype<a class="anchor" aria-label="anchor" href="#arg-min-per-genotype"></a></dt>
<dd><p>Minimum number of individuals required in each
genotype group (homozygous reference, heterozygous, homozygous alternate)</p></dd>


<dt id="arg-genome">genome<a class="anchor" aria-label="anchor" href="#arg-genome"></a></dt>
<dd><p>Genome identifier (e.g., "hg19", "hg38")</p></dd>


<dt id="arg-regions">regions<a class="anchor" aria-label="anchor" href="#arg-regions"></a></dt>
<dd><p>Obligatory GRanges object to restrict to specific genomic regions</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Logical, print progress messages</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A filtered VCF object</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">filtvcf</span> <span class="op">&lt;-</span> <span class="fu">filterVcfByGenotypeCounts</span><span class="op">(</span></span></span>
<span class="r-in"><span>  vcf_file <span class="op">=</span> <span class="fu"><a href="demo_vcf.html">demo_vcf</a></span><span class="op">(</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"HG00096"</span>, <span class="st">"HG00100"</span>, <span class="st">"HG00105"</span>, <span class="st">"HG00108"</span>, <span class="st">"HG00110"</span>, <span class="st">"HG00113"</span>, </span></span>
<span class="r-in"><span>     <span class="st">"HG00114"</span>, <span class="st">"HG00121"</span>, <span class="st">"HG00122"</span>, <span class="st">"HG00127"</span>, <span class="st">"HG00130"</span>, <span class="st">"HG00132"</span>, </span></span>
<span class="r-in"><span>     <span class="st">"HG00141"</span>, <span class="st">"HG00142"</span>, <span class="st">"HG00146"</span>, <span class="st">"HG00148"</span>, <span class="st">"HG00151"</span>, <span class="st">"HG00160"</span>, </span></span>
<span class="r-in"><span>     <span class="st">"HG00177"</span>, <span class="st">"HG00179"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  min_per_genotype <span class="op">=</span> <span class="fl">2</span>,</span></span>
<span class="r-in"><span>  regions <span class="op">=</span> <span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"19:1-50000000"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  genome <span class="op">=</span> <span class="st">"hg38"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Total samples in VCF: 731</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analyzing 20 samples</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading VCF...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Read 8465 variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Counting genotypes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Variants passing filter: 2922 / 8465</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Mean counts - HomRef: 10.8, Het: 6, HomAlt: 3.1</span>
<span class="r-in"><span><span class="va">filtvcf</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: CollapsedVCF </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 2922 20 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowRanges(vcf):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   GRanges with 5 metadata columns: paramRangeID, REF, ALT, QUAL, FILTER</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> info(vcf):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   DataFrame with 27 columns: CIEND, CIPOS, CS, END, IMPRECISE, MC, MEINFO, M...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> info(header(vcf)):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  Number Type    Description                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    CIEND         2      Integer Confidence interval around END for imprecise...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    CIPOS         2      Integer Confidence interval around POS for imprecise...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    CS            1      String  Source call set.                               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    END           1      Integer End coordinate of this variant                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    IMPRECISE     0      Flag    Imprecise structural variation                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    MC            .      String  Merged calls.                                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    MEINFO        4      String  Mobile element info of the form NAME,START,E...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    MEND          1      Integer Mitochondrial end coordinate of inserted seq...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    MLEN          1      Integer Estimated length of mitochondrial insert       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    MSTART        1      Integer Mitochondrial start coordinate of inserted s...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    SVLEN         .      Integer SV length. It is only calculated for structu...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    SVTYPE        1      String  Type of structural variant                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    TSD           1      String  Precise Target Site Duplication for bases, i...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    AC            A      Integer Total number of alternate alleles in called ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    AF            A      Float   Estimated allele frequency in the range (0,1)  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    NS            1      Integer Number of samples with data                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    AN            1      Integer Total number of alleles in called genotypes    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    EAS_AF        A      Float   Allele frequency in the EAS populations calc...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    EUR_AF        A      Float   Allele frequency in the EUR populations calc...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    AFR_AF        A      Float   Allele frequency in the AFR populations calc...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    AMR_AF        A      Float   Allele frequency in the AMR populations calc...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    SAS_AF        A      Float   Allele frequency in the SAS populations calc...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    DP            1      Integer Total read depth; only low coverage data wer...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    AA            1      String  Ancestral Allele. Format: AA|REF|ALT|IndelTy...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    VT            .      String  indicates what type of variant the line repr...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    EX_TARGET     0      Flag    indicates whether a variant is within the ex...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    MULTI_ALLELIC 0      Flag    indicates whether a site is multi-allelic      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> geno(vcf):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   List of length 1: GT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> geno(header(vcf)):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       Number Type   Description</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    GT 1      String Genotype   </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Vince Carey.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

