<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ biocXqtl</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">biocXqtl</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.6</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/biocXqtl.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/vjcitn/biocXqtl/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="XqtlExperiment-class.html">XqtlExperiment-class</a></code>

        </dt>
        <dd>extend RangedSummarizedExperiment to include genotype calls in a GRanges</dd>

        <dt>

          <code><a href="XqtlExperiment.html">XqtlExperiment()</a></code>

        </dt>
        <dd>XqtlExperiment constructor</dd>

        <dt>

          <code><a href="add_calls_from_vcf.html">add_calls_from_vcf()</a></code>

        </dt>
        <dd>add minor allele counts from a VCF to SummarizedExperiment with common samples</dd>

        <dt>

          <code><a href="bind_Zs.html">bind_Zs()</a></code>

        </dt>
        <dd>bind a matrix of Z statistics created with getzs to the rowRanges of a SummarizedExperiment</dd>

        <dt>

          <code><a href="countGenotypes.html">countGenotypes()</a></code>

        </dt>
        <dd>Count genotypes across samples for each variant</dd>

        <dt>

          <code><a href="countGenotypesVectorized.html">countGenotypesVectorized()</a></code>

        </dt>
        <dd>More efficient vectorized version for large datasets</dd>

        <dt>

          <code><a href="demo_vcf.html">demo_vcf()</a></code>

        </dt>
        <dd>provide path to a demonstration VCF file</dd>

        <dt>

          <code><a href="fastLmMany_R.html">fastLmMany_R()</a></code>

        </dt>
        <dd>Fit multiple linear models with common design matrix</dd>

        <dt>

          <code><a href="filterCalls.html">filterCalls()</a></code>

        </dt>
        <dd>filter the call component of an XqtlExperiment</dd>

        <dt>

          <code><a href="filterVcfByGenotypeCounts.html">filterVcfByGenotypeCounts()</a></code>

        </dt>
        <dd>Filter VCF to variants with minimum counts in each genotype group</dd>

        <dt>

          <code><a href="filterVcfMultipleCriteria.html">filterVcfMultipleCriteria()</a></code>

        </dt>
        <dd>Filter VCF with multiple criteria</dd>

        <dt>

          <code><a href="getCalls.html">getCalls()</a></code>

        </dt>
        <dd>getter for genotype calls</dd>

        <dt>

          <code><a href="getGenotypeSummary.html">getGenotypeSummary()</a></code>

        </dt>
        <dd>Get genotype summary statistics</dd>

        <dt>

          <code><a href="getzs.html">getzs()</a></code>

        </dt>
        <dd>create a matrix of Z statistics for an additive effect of rare allele count in calls on molecular phenotype in molec</dd>

        <dt>

          <code><a href="geuv19.html">geuv19</a></code>

        </dt>
        <dd>a RangedSummarizedExperiment with data from GeuvadisTranscriptExpr</dd>

        <dt>

          <code><a href="geuv19_samples.html">geuv19_samples</a></code>

        </dt>
        <dd>data frame with sample information for geuv19</dd>

        <dt>

          <code><a href="geuv19xse.html">geuv19xse</a></code>

        </dt>
        <dd>a RangedSummarizedExperiment with data from GeuvadisTranscriptExpr, in XqtlExperiment format</dd>

        <dt>

          <code><a href="maf.html">maf()</a></code>

        </dt>
        <dd>compute putative minor allele frequency for XqtlExperiment</dd>

        <dt>

          <code><a href="mageSE_19.html">mageSE_19</a></code>

        </dt>
        <dd>SummarizedExperiment from MAGE</dd>

        <dt>

          <code><a href="minorAlleleCounts.html">minorAlleleCounts()</a></code>

        </dt>
        <dd>produce data.frame with minor allele counts in a region, from VCF</dd>

        <dt>

          <code><a href="show-XqtlExperiment-method.html">show(<i>&lt;XqtlExperiment&gt;</i>)</a></code>

        </dt>
        <dd>present concise view of XqtlExperiment</dd>

        <dt>

          <code><a href="viz_stats.html">viz_stats()</a></code>

        </dt>
        <dd>produce a plotly display of statistics of xQTL association</dd>

        <dt>

          <code><a href="zs4manyYs.html">zs4manyYs()</a></code>

        </dt>
        <dd>use procedure tailored to multiple responses for a fixed design matrix</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Vince Carey.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

